lib/semantics/check-omp-structure.cc

```diff
@@ -82,10 +82,55 @@ void OmpStructureChecker::Enter(const parser::OpenMPLoopConstruct &x) {
   // push a context even in the error case
   OmpContext ct{dir.source};
   ompContext_.push_back(ct);
+  checkState_.justSawLoopConstruct = true;
+}
+
+void OmpStructureChecker::Enter(const parser::ExecutableConstruct &x) {
+  if (checkState_.justSawLoopConstruct) {
+    if (const auto *doDir{
+            std::get_if<common::Indirection<parser::DoConstruct>>(&x.u)}) {
+      // save associated DoConstruct pointer
+      SetContextDo(doDir->value());
+      checkState_.inOmpLoopRegion = true;
+    } else {
+      context_.Say(GetContext().directiveSource,
+          "do-loop is expected after the %s directive"_err_en_US,
+          ContextDirectiveAsFortran());
+      ompContext_.pop_back();
+    }
+    checkState_.justSawLoopConstruct = false;
+  }
+
+  const auto *endDoDir{
+      std::get_if<common::Indirection<parser::OpenMPEndLoopDirective>>(&x.u)};
+  if (!endDoDir && checkState_.allowOptionalEndDirective) {
+    // END directive is not present, loop construct is complete
+    checkState_.allowOptionalEndDirective = false;
+  }
 }
 
-void OmpStructureChecker::Leave(const parser::OpenMPLoopConstruct &) {
-  ompContext_.pop_back();
+void OmpStructureChecker::Leave(const parser::ExecutableConstruct &x) {
```

Comment:
Does this work if you see an `ExecutionPartConstruct` that is not an `ExecutableConstruct`?
For example, what if a loop directive is followed by a FORMAT statement?

Reply:
Sorry, I should've noticed this, I will fix it.